Mesh_3/include/CGAL/Mesh_3/generate_label_weights.h

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) 2021 GeometryFactory
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+//
+// Author(s)     : Laurent Rineau, Jane Tournois
+
+#ifndef CGAL_MESH_3_GENERATE_LABEL_WEIGHTS_H
+#define CGAL_MESH_3_GENERATE_LABEL_WEIGHTS_H
+
+#include <CGAL/license/Mesh_3.h>
+
+#include <CGAL/Image_3.h>
+#include <CGAL/ImageIO.h>
+
+#include <itkImage.h>
+#include <itkImageDuplicator.h>
+#include <itkBinaryThresholdImageFilter.h>
+#include <itkRecursiveGaussianImageFilter.h>
+#include <itkMaximumImageFilter.h>
+
+#include <iostream>
+#include <vector>
+#include <set>
+#include <type_traits>
+
+namespace CGAL {
+namespace Mesh_3 {
+namespace internal {
+
+template<typename Image_word_type, typename LabelsSet>
+void convert_image_3_to_itk(const CGAL::Image_3& image,
+                            itk::Image<Image_word_type, 3>* const itk_img,
+                            LabelsSet& labels)
+{
+  using ImageType = itk::Image<Image_word_type, 3/*Dimension*/>;
+
+  typename ImageType::SpacingType spacing;
+  spacing[0] = image.vx();
+  spacing[1] = image.vy();
+  spacing[2] = image.vz();
+  itk_img->SetSpacing(spacing);
+
+  typename ImageType::PointType origin;
+  origin[0] = image.tx();
+  origin[1] = image.ty();
+  origin[2] = image.tz();
+  itk_img->SetOrigin(origin);
+
+  typename ImageType::IndexType  corner = {{0, 0, 0 }};
+  typename ImageType::SizeType   size = {{image.xdim(), image.ydim(), image.zdim()}};
+  typename ImageType::RegionType region(corner, size);
+  itk_img->SetRegions(region);
+
+//  itk_img->Allocate(image.xdim() * image.ydim() * image.zdim());
+  itk_img->Allocate();
+
+  using Index = itk::Index<3>::IndexValueType;
+  for (std::size_t i = 0; i < image.xdim(); ++i)
+  {
+    for (std::size_t j = 0; j < image.ydim(); ++j)
+    {
+      for (std::size_t k = 0; k < image.zdim(); ++k)
+      {
+        typename ImageType::IndexType index = {(Index)i, (Index)j, (Index)k};
+        const Image_word_type label = image.value(i, j, k);
+        labels.insert(label);
+        itk_img->SetPixel(index, label);
+      }
+    }
+  }
+}
```

Comment:
I find that particularly non-efficient.

The [documentation of `itk::Image`](https://itk.org/Doxygen/html/classitk_1_1Image.html#aa6ce68e3fc18faf29b2cbb8dce2187b9) says:
> The data in an image is arranged in a 1D array as `[][][][slice][row][col]` with the column index varying most rapidly.

If that is the same convention in `CGAL::Image_3`, then we can just call `std::copy` from the data of `image` to `itk_img->GetBufferPointer()`. As for `labels`: use `labels.insert(begin, end)` instead of writing a custom loop.

Reply:
```c++
auto begin = static_cast<Image_word_type*>(image.data());
auto end = begin + image.size();
labels.insert(begin, end);
```
but not in a loop.